### PR TITLE
add a script for use when new kde release is made

### PR DESCRIPTION
**Pull request by jriddell**

### CODE REVIEW

File: jenkins_trigger_kde_release_build.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+  opts.on('-r TYPE', '--release=TYPE', 'Which release is new') do |r|
+    release = r
+  end
+end.parse!
+
+releases = %w[plasma applications frameworks]
+
+unless releases.include? release
+  puts '-r TYPE must be one of plasma, applications, frameworks'
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: That's not true though is it? It needs to be a method of `KDEProjectsComponent`. Hardcoding this list doesn't serve much purpose.

```
unless KDEProjectsComponent.respond_to?(release)
   abort "#{release} isn't a supported type"
end
```

File: jenkins_trigger_kde_release_build.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  exit
+end
+
+@log = Logger.new(STDOUT).tap do |l|
+  l.progname = 'trigger_kde_release_build'
+  l.level = Logger::INFO
+end
+
+packages = KDEProjectsComponent.public_send(release)
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: Bad var name. That's a list of project names, not packages.

**Comment by jriddell** [comment]
Issues should be resolved

File: lib/jenkins/retry.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+require_relative '../../ci-tooling/lib/jenkins'
+
+module Jenkins
+
```
Comment (apachelogger) [inline lib/jenkins/retry.rb]: excess newline

File: lib/jenkins/retry.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  def retry_jobs(job_name_queue, exclusion_states=%w[success unstable], strict_mode=false)
```
Comment (apachelogger) [inline lib/jenkins/retry.rb]: method too long

all arguments but the queue should be keyword arguments

(also strict_mode is a really bad name, could be called foobar and it'd be just as indicative of what it does)

File: jenkins_trigger_kde_release_build.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+job_name_queue = Queue.new
+job_names = Jenkins.job.list_all
+projects.each do |x|
+  job_name_queue << "watcher_release_kde_#{x}" if job_names.include? "watcher_release_kde_#{x}"
+end
+
+@log.info 'Setting system into maintenance mode.'
+Jenkins.system.quiet_down
+
+Jenkins.retry(job_name_queue, exclusion_states, strict_mode)
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: This should be a Retrier class, the Retirer class is configured and then run.

```
retrier = Jenkins::Retrier.new(exclude_states: %w[foo bar], strict_mode: false)\
retrier.retry(queue)
```

File: lib/kdeproject_component.rb [hunk 6]
```diff
@@ -40,6 +40,10 @@ def plasma
       @plasma ||= to_names(projects('kde/workspace'))
     end
 
+    def respond_to?(release)
```
Comment (apachelogger) [inline lib/kdeproject_component.rb]: What is this about?

This is already a core feature of all objects...
https://ruby-doc.org/core-2.5.0/Object.html#method-i-respond_to-3F

File: lib/jenkins/retry.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+require_relative '../../ci-tooling/lib/jenkins'
+
```
Comment (apachelogger) [inline lib/jenkins/retry.rb]: missing a whole bunch of requires

File: jenkins_trigger_kde_release_build.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  END_OF_HELP
+
+  opts.on('-r TYPE', '--release=TYPE', 'Which release is new') do |r|
+    release = r
+  end
+end.parse!
+
+releases = %w[plasma applications frameworks]
+
+unless KDEProjectsComponent.respond_to? release
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: parenthesis around method arguments please

File: jenkins_trigger_kde_release_build.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  puts '-r TYPE must be one of plasma, applications, frameworks'
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: puts & exit = abort https://ruby-doc.org/core-2.5.0/Kernel.html#method-i-abort

File: lib/jenkins/retry.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+module Jenkins
+
+  def retry_jobs(job_name_queue, exclusion_states=%w[success unstable], strict_mode=false)
+    BlockingThreadPool.run do
+      until job_name_queue.empty?
+        name = job_name_queue.pop(true)
+        Retry.retry_it(times: 5) do
+          status = Jenkins.job.status(name)
+          queued = Jenkins.client.queue.list.include?(name)
+          @log.info "#{name} | status - #{status} | queued - #{queued}"
```
Comment (apachelogger) [inline lib/jenkins/retry.rb]: undefined variable

File: jenkins_trigger_kde_release_build.rb [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+require 'logger'
+require 'logger/colors'
+require 'optparse'
+
+require_relative 'ci-tooling/lib/jenkins'
+require_relative 'ci-tooling/lib/retry'
+require_relative 'ci-tooling/lib/thread_pool'
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: not used (same in other script)

File: jenkins_trigger_kde_release_build.rb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+require 'logger'
+require 'logger/colors'
+require 'optparse'
+
+require_relative 'ci-tooling/lib/jenkins'
+require_relative 'ci-tooling/lib/retry'
```
Comment (apachelogger) [inline jenkins_trigger_kde_release_build.rb]: not used (same in other script)

**Pull request closed without merging**